metals/src/main/scala/scala/meta/internal/metals/FormattingProvider.scala

```diff
@@ -215,6 +223,155 @@ final class FormattingProvider(
     } else Future.successful(false)
   }
 
+  private def initialConfig(): String = {
+    val version = BuildInfo.scalafmtVersion
+    val initialConfig =
+      ScalafmtConfig.empty
+        .copy(version = Some(SemVer.Version.fromString(version)))
+
+    val versionText = s"""version = "${BuildInfo.scalafmtVersion}""""
+    inspectDialectRewrite(initialConfig) match {
+      case Some(rewrite) => rewrite.rewrite(versionText)
+      case None => versionText
+    }
+  }
+
+  private def inspectDialectRewrite(
+      config: ScalafmtConfig
+  ): Option[DialectRewrite] = {
+
+    import ScalafmtDialect.ord
+
+    def hasCorrectDialectOverride(
+        files: List[AbsolutePath],
+        dialect: ScalafmtDialect
+    ): Boolean = {
+      config.fileOverrides.nonEmpty &&
+      files.forall(p => config.overrideFor(p).exists(d => ord.gteq(dialect, d)))
+    }
+
+    def inferDialectForSourceItem(
+        p: AbsolutePath,
+        buildTargetIds: List[BuildTargetIdentifier],
+        default: ScalafmtDialect
+    ): Option[ScalafmtDialect] = {
+      val scalaFiles =
```

Comment:
I would defer calculating `scalaFiles` until `hasCorrectDialectOverride`:

```scala
    def hasCorrectDialectOverride(
        files: List[AbsolutePath],
        dialect: ScalafmtDialect
    ): Boolean = {
     lazy val scalaFiles = ... // calculate them
      config.fileOverrides.nonEmpty &&
      scalaFiles.nonEmpty &&
      scalaFiles.forall(p => config.overrideFor(p).exists(d => ord.gteq(dialect, d)))
    }
```

Otherwise, I am afraid we are adding a whole lot of effort whenever we index the workspace. And this way if everything is set ok, we will not need to calculate anything.